datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -254,67 +298,103 @@ void send(DatastreamTask task, ProducerRecord<K, V> producerRecord, Callback onC
       }
     }
   }
 
   @VisibleForTesting
   void shutdownProducer() {
     Producer<K, V> producer;
-    synchronized (_producerLock) {
+    try {
+      _producerLock.lock();
+      // if there is no producer or the producer close is in progress, return.
+      if (_kafkaProducer == null || _closeInProgress) {
+        return;
+      }
       producer = _kafkaProducer;
+      _closeInProgress = true;
       // Nullify first to prevent subsequent send() to use
       // the current producer which is being shutdown.
       _kafkaProducer = null;
-    }
 
-    // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
-    // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
-    // thread
-    if (producer != null) {
+      // This may be called from the send callback. The callbacks are called from the sender thread, and must complete
+      // quickly to avoid delaying/blocking the sender thread. Thus schedule the actual producer.close() on a separate
+      // thread
       _producerCloseExecutorService.submit(() -> {
         _log.info("KafkaProducerWrapper: Closing the Kafka Producer");
-        producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
-        NUM_PRODUCERS.decrementAndGet();
-        _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        try {
+          producer.close(CLOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+          NUM_PRODUCERS.decrementAndGet();
+          _log.info("KafkaProducerWrapper: Kafka Producer is closed");
+        } finally {
+          markProducerCloseComplete();
+        }
       });
+    } finally {
+      _producerLock.unlock();
     }
   }
 
-  private DatastreamRuntimeException generateSendFailure(Exception exception) {
+  private void markProducerCloseComplete() {
+    try {
+      _producerLock.lock();
+      _closeInProgress = false;
+      _waitOnProducerClose.signalAll();
+    } finally {
+      _producerLock.unlock();
+    }
+  }
+
+  private DatastreamRuntimeException generateSendFailure(Exception exception, DatastreamTask task) {
     _dynamicMetricsManager.createOrUpdateMeter(_metricsNamesPrefix, AGGREGATE, PRODUCER_ERROR, 1);
     if (exception instanceof IllegalStateException) {
       _log.warn("Send failed transiently with exception: ", exception);
       return new DatastreamTransientException(exception);
     } else {
       _log.warn("Send failed with a non-transient exception. Shutting down producer, exception: ", exception);
-      shutdownProducer();
+      if (_tasks.contains(task)) {
+        shutdownProducer();
+      }
       return new DatastreamRuntimeException(exception);
     }
   }
 
   void flush() {
-    synchronized (_producerLock) {
+    Producer<K, V> producer;
+    try {
+      _producerLock.lock();
+      producer = _kafkaProducer;
+    } finally {
+      _producerLock.unlock();
+    }
+
+    try {
```

Comment:
sorry I might be missing the reason the code looks like this but should it read
```
    if (producer != null) {
       try {
          producer.flush (...)
       } catch (..) {
         ...
       }
  } ```

Reply:
done.